### PR TITLE
STRG fixes for non US versions

### PR DESCRIPTION
**Pull request by UltiNaruto**
# Seed hash not showing on PAL main menu

Seed hash is not shown on PAL main menu when starting the game for the first time in a non english language. This will remain the thing until the save file is deleted on memory card (not in main menu). This commit fixes the above.

Link to bug report: [#bug-reports -> Seed doesn't display on main menu on PAL](https://discord.com/channels/914291389293027329/1175534362914193419)

# Credits not showing on PAL version

Credits is now properly patched on every languages.

# Credits being cut on JP version

Credits were cut at Flamethrower due to line-extra-space being set too high.

# Fixed font size on JP version

Most of the edited text were using a way bigger font. With this PR, it's fixed.

# STRG helper functions

Strg.add_strings(strings: Vec<String>) calls Strg.add_strings_2(strings, all_languages) which adds all the strings at the end of the tables.
Same for Strg.edit_strings((from, to): (String, String)) which replaces the `from` string with `to` string

Any other changes are just trimming whitespaces. It's done automatically via Notepad++ with Alt+Shift+S to save the file.

### CODE REVIEW

**Comment by toasterparty** [comment]
This feels like something where we should just have a list of supported language FourCC and always patch all of them where they exist. Without such a solution, your fix here needs to be copy-pasted for all other localized strings we modify like the credits patch.

Could use like a helper function or whatever.

**Comment by UltiNaruto** [comment]
then we should have a helper function that passes the version as argument

**Comment by blake-spangenberg** [comment]
I don't think you need to change any behavior based on version.

It'd be something like

```rs
for lang in LANGUAGES {
    let strings = res.kind.as_strg_mut().unwrap()
        .string_tables
        .as_mut_vec()
        .iter_mut()
        .find(|table| table.lang == lang.into());
    if strings.is_none() {continue;}
    let strings = strings.unwrap().strings.as_mut_vec();
    strings_pal.push(msg).into());
}
```

**Comment by blake-spangenberg** [comment]
Or you might be able to just remove the `.find` alltogether and just do `for table in string_tables`

**Comment by toasterparty** [comment]
I don't think you need to change any behavior based on version.

It'd be something like

```rs
for lang in LANGUAGES {
    let table= res.kind.as_strg_mut().unwrap()
        .string_tables
        .as_mut_vec()
        .iter_mut()
        .find(|table| table.lang == lang.into());
    if strings.is_none() {continue;}
    let strings = table.unwrap().strings.as_mut_vec();
    // do the thing
}
```

Alternatively, you might even be able to get away with

```rs
let tables = res.kind
    .as_strg_mut()
    .unwrap()
    .string_tables
    .as_mut_vec();
for table in tables {
    let strings = table.strings.as_mut_vec();
    // do the thing
}
```

If you are confident that the only reason tables are separated in the first place is for internationalization reasons

**Comment by UltiNaruto** [comment]
Edited the PR and added helper functions to Strg

File: structs/src/strg.rs [hunk 1]
```diff
@@ -37,6 +37,68 @@ pub struct Strg<'r>
 
 impl<'r> Strg<'r>
 {
+    fn is_jap_version(languages: Vec<FourCC>) -> bool
+    {
+        languages.len() == 2 && languages.iter().any(|lang| lang.as_bytes() == b"ENGL") && languages.iter().any(|lang| lang.as_bytes() == b"JAPN")
+    }
+
+    pub fn add_strings_2(self: &mut Self, strings: Vec<String>, languages: Vec<FourCC>)
```
Comment (toasterparty) [inline structs/src/strg.rs]: Please make a single function with `languages: Option<Vec<FourCC>>`, where `None` defaults to 

```rs
vec![
        FourCC::from(b"ENGL"),
        FourCC::from(b"DUTC"),
        FourCC::from(b"FREN"),
        FourCC::from(b"GERM"),
        FourCC::from(b"ITAL"),
        FourCC::from(b"JAPN"),
        FourCC::from(b"SPAN"),
    ]
```

File: structs/src/strg.rs [hunk 2]
```diff
@@ -37,6 +37,68 @@ pub struct Strg<'r>
...
+            FourCC::from(b"FREN"),
+            FourCC::from(b"GERM"),
+            FourCC::from(b"ITAL"),
+            FourCC::from(b"JAPN"),
+            FourCC::from(b"SPAN"),
+        ];
+        self.add_strings_2(strings, supported_languages);
+    }
+
+    pub fn edit_strings_2(self: &mut Self, (from, to): (String, String), languages: Vec<FourCC>)
```
Comment (toasterparty) [inline structs/src/strg.rs]: Same as above. You should also use a shared constant vec for the supported languages. If the compiler complains, you can use [lazy_static](https://docs.rs/lazy_static/latest/lazy_static/)

File: structs/src/strg.rs [hunk 3]
```diff
@@ -37,6 +37,68 @@ pub struct Strg<'r>
 
 impl<'r> Strg<'r>
 {
+    fn is_jap_version(languages: Vec<FourCC>) -> bool
```
Comment (toasterparty) [inline structs/src/strg.rs]: This abbreviation of "Japanese" is actually a racial slur whose history can be traced back to war crimes

jpn/jp/j are all fine alternatives

**Review by toasterparty — CHANGES_REQUESTED** [review]

File: structs/src/strg.rs (hunk 3 above)
Comment (UltiNaruto) [reply structs/src/strg.rs]: I didn't know that

Comment (UltiNaruto) [reply structs/src/strg.rs]: btw that means in the dol symbol table we should change to jpn instead of jap?

**Comment by UltiNaruto** [comment]
Updated the PR with the requested changes. I went for a different way for the optional argument. I used an enum instead of Option which feels much better to me.

File: structs/src/strg.rs (hunk 3 above)
Comment (toasterparty) [reply structs/src/strg.rs]: If it's something retro did no, if it's something we did, then probably yeah...

Comment (UltiNaruto) [reply structs/src/strg.rs]: I mean check in generated/dol_symbol_table folder

**Pull request merged**